### PR TITLE
T40660 workflows: enable python-lint-annotate

### PR DESCRIPTION
**Pull request by JenySadadia**
Fixes https://github.com/kernelci/kernelci-api/issues/212

Add a job 'lint' using python-lint-annotate to run pylint with GitHub inline annotations.

Tested with https://github.com/kernelci/kernelci-api/pull/217

### CODE REVIEW

**Comment by JenySadadia** [comment]
Hello @gctucker 
I am checking why the `lint` job isn't failing even if it finds issues with `models.py`.

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -39,23 +39,42 @@ jobs:
...
-
-      - name: Run pylint
-        run: |
-          pylint api.auth
-          pylint api.db
-          pylint -d R0903 api.models
-          pylint api.pubsub
-          pylint test
-          pylint api.main
-
```
Comment (gctucker) [inline .github/workflows/main.yml:54]: We need to keep this to report errors, the annotations only add annotations.

**Review by gctucker — CHANGES_REQUESTED** [review]

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -39,23 +39,42 @@ jobs:
...
+        with:
+          python-root-list: ${{ env.CHANGED_FILES }}
+          use-black: false
+          use-flake8: false
+          use-isort: false
+          use-mypy: false
+          use-pycodestyle: true
+          use-pydocstyle: false
+          use-vulture: false
+          python-version: "3.10"
```
Comment (gctucker) [inline .github/workflows/main.yml:80]: I believe this should be dropped as the setup-python step already sets it up with a single version.  IIRC this option is only useful when several Python versions are installed.

File: .github/workflows/main.yml (hunk 1 above)
Comment (JenySadadia) [reply .github/workflows/main.yml:54]: That's what I mentioned in the earlier comment. I thought there should be a way to fail `lint` job.

Comment (JenySadadia) [reply .github/workflows/main.yml:54]: Anyway, I'll revert this.

Comment (gctucker) [reply .github/workflows/main.yml]: Yes I had a look at this when enabling it on kernelci-core and I didn't see an obvious way to make the annotation action fail the check if some linters reported any errors.  So for now we need to keep both actions.

File: .github/workflows/main.yml (hunk 2 above)
Comment (JenySadadia) [reply .github/workflows/main.yml:93]: setup-python step explicitly declares python version 3.10 to be used.
But somehow the lint job uses python version 3.8 from actions within `marian-code/python-lint-annotate@v3`.
I added `python-version` to correct it.

Comment (gctucker) [reply .github/workflows/main.yml:93]: Ah I guess `ubuntu-22.04` comes with Python 3.8 already installed then.  In this case I guess we need to keep the explicit `python-version` option and I should do the same with `kernelci-core`.

File: .github/workflows/main.yml (hunk 1 above)
Comment (JenySadadia) [reply .github/workflows/main.yml]: Oh, okay.

File: .github/workflows/main.yml (hunk 2 above)
Comment (JenySadadia) [reply .github/workflows/main.yml:93]: Yes.

Comment (gctucker) [reply .github/workflows/main.yml:93]: Actually, I double checked and `ubuntu-22.04` comes with Python 3.10.  The `marian-code` action runs its own `setup-python` step with 3.8 by default:
```
Run actions/setup-python@v2
  with:
    python-version: 3.8
    architecture: x64
    cache: pip
    token: ***
  env:
    pythonLocation: /opt/hostedtoolcache/Python/3.10.10/x64
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.10.10/x64/lib
    CHANGED_FILES: kci kernelci/config/api.py
```
So we can just drop the first `setup-python` step in `main.yml` and rely only on the `python-version` in the annotation action.

Comment (JenySadadia) [reply .github/workflows/main.yml:93]: Okay. Thanks for confirming.

Comment (JenySadadia) [reply .github/workflows/main.yml:88]: Also, why the value 32 is used in `fetch-depth: 32` in `kernelci-core`?

Comment (gctucker) [reply .github/workflows/main.yml:88]: Because we never have PRs with more than 32 commits, in practice.

Comment (gctucker) [reply .github/workflows/main.yml:88]: I've added a commit to this PR to use the `python-version` attribute in the annotation action:
https://github.com/kernelci/kernelci-core/pull/1748

**Comment by gctucker** [comment]
I see the linter annotation action is reporting issues with importing `bson` and `pydantic` packages.  It seems like the `requirements.txt` were not installed.  Now I wonder how that's working for `kernelci-core`.

**Comment by gctucker** [comment]
> I see the linter annotation action is reporting issues with importing `bson` and `pydantic` packages. It seems like the `requirements.txt` were not installed. Now I wonder how that's working for `kernelci-core`.

OK I've reworked this accordingly to install `requirements.txt` in the linter annotations action and kept the `setup-python` step too:
https://github.com/kernelci/kernelci-core/pull/1748

**Comment by JenySadadia** [comment]
> I see the linter annotation action is reporting issues with importing `bson` and `pydantic` packages. It seems like the `requirements.txt` were not installed. Now I wonder how that's working for `kernelci-core`.

Okay. I need to fix this here as well.

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -59,3 +59,41 @@ jobs:
...
+
+  lint:
+    runs-on: ubuntu-22.04
+    name: Lint
+
+    steps:
+      - name: Check out source code
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0  # This is necessary to get the commits
```
Comment (gctucker) [inline .github/workflows/main.yml:71]: The issue with `0` is you get the full history, which can take a lot of time and isn't useful past the commits in the PR.

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -59,3 +59,41 @@ jobs:
...
+      - if: env.CHANGED_FILES
+        name: Set up Python
+        uses: actions/setup-python@master
+        with:
+          python-version: "3.10"
+
+      - if: env.CHANGED_FILES
+        name: Install Python packages
+        run: |
+          pip install -r requirements-dev.txt
```
Comment (gctucker) [inline .github/workflows/main.yml:86]: I think it should be `requirements.txt` rather than the `-dev` one.  The linter tools etc. are installed later.

**Review by gctucker — COMMENTED** [review]
I think it would be worth creating a temporary draft PR with some linter errors to check they're working as expected.  See what I did for kernelci-core:
https://github.com/kernelci/kernelci-core/pull/1749

If you have other means of checking it's working as expected then fine, but it's hard to tell at the moment.

File: .github/workflows/main.yml (hunk 4 above)
Comment (JenySadadia) [reply .github/workflows/main.yml:86]: Files under `test` are using packages from `requirements-dev.txt`. Hence need to install them as well.

**Comment by JenySadadia** [comment]
> I think it would be worth creating a temporary draft PR with some linter errors to check they're working as expected. See what I did for kernelci-core: [kernelci/kernelci-core#1749](https://github.com/kernelci/kernelci-core/pull/1749)
> 
> If you have other means of checking it's working as expected then fine, but it's hard to tell at the moment.

For the sake of testing, I added a few temporary changes to this PR itself and then discarded it after testing.

**Comment by gctucker** [comment]
> 

OK great.  I can't use that to review the PR though so I'll just have to assume it's working as expected ;)

File: .github/workflows/main.yml (hunk 4 above)
Comment (gctucker) [reply .github/workflows/main.yml]: Ah OK, that's a difference with `kernelci-core` where the main `requirements.txt` file has everything to run unit tests as well e.g. `pytest`.  I'm not sure what's the general convention for things that should go in `-dev.txt`.

**Comment by JenySadadia** [comment]
> > 
> 
> OK great. I can't use that to review the PR though so I'll just have to assume it's working as expected ;)

Created https://github.com/kernelci/kernelci-api/pull/217/files for testing.

File: .github/workflows/main.yml (hunk 4 above)
Comment (gctucker) [reply .github/workflows/main.yml]: So `requirements.txt` has `pylint==2.12.2` here.  Which version of pylint is used by the annotation action?

Comment (gctucker) [reply .github/workflows/main.yml]: Interestingly, the annotation action installs `pylint==2.12.1`.

Comment (gctucker) [reply .github/workflows/main.yml]: Not sure how to debug that.  For now it shouldn't matter as they're very close versions but if for any reason we start running different pylint versions or other linter tools then we could start having different errors being reported which can be inconvenient.

Maybe `requirements-dev.txt` could be split into one file with the unit tests requirements and another one with the linters?  Say `requirements-tests.txt` which would include `requirements.txt` and add things like `pytest` and `fakeredis`, then `requirements-dev.txt` which would include `requirements-tests.txt` and add `pylint` and `pycodestyle` on top.  Then the annotations would be using `requirements-tests.txt`.

Comment (JenySadadia) [reply .github/workflows/main.yml]: The general convention is requirements for testing and development go to `requirements-dev.txt`. Still, there are no hard rules for that.

File: .github/workflows/main.yml (hunk 3 above)
Comment (JenySadadia) [reply .github/workflows/main.yml:71]: I need to fix this.

File: .github/workflows/main.yml (hunk 4 above)
Comment (JenySadadia) [reply .github/workflows/main.yml]: Yes, we need to resolve the issues with version conflicts. 
It is a good idea to split requirements into multiple files.
How about adding `pylint` and `pycodestyle` to `requirements-linter.txt` file? Then the linter requirements file will include `requirements-tests.txt`.

File: api/models.py [hunk 5]
```diff
@@ -4,6 +4,8 @@
 # Author: Guillaume Tucker <guillaume.tucker@collabora.com>
 # Author: Jeny Sadadia <jeny.sadadia@collabora.com>
 
+# pylint: disable=too-few-public-methods
```
Comment (gctucker) [inline api/models.py:7]: It might be good to add a comment about why it's OK to disable this check (basically, because models don't have methods...).

File: .github/workflows/main.yml [hunk 6]
```diff
@@ -67,11 +67,17 @@ jobs:
     steps:
       - name: Check out source code
         uses: actions/checkout@v3
+        with:
+          fetch-depth: 32  # This is necessary to get the commits
```
Comment (gctucker) [inline .github/workflows/main.yml:71]: This is not related to looking for changed files.  Maybe it should just be part of the first commit that adds the linter job?

File: .github/workflows/main.yml [hunk 7]
```diff
@@ -74,6 +74,17 @@ jobs:
         run: >
           echo "CHANGED_FILES=$(echo $(git diff --name-only ${{ github.event.pull_request.base.sha }}..${{ github.event.pull_request.head.sha }} -- | grep \.py))" >> $GITHUB_ENV
 
+      - if: env.CHANGED_FILES
```
Comment (gctucker) [inline .github/workflows/main.yml:77]: I found the commit subject a bit misleading, it looked like some fixes in the Python code to fix import errors reported by pylint.  Maybe something like "install requirements" would be clearer.

**Review by gctucker — COMMENTED** [review]
I think we can ignore the fact that pylint and pycodestyle are being installed twice for now as it's not causing any issue in practice.  So we should be able to merge this PR with the regular `requirements-dev.txt` and then rework this somehow.

**Comment by JenySadadia** [comment]
> I think we can ignore the fact that pylint and pycodestyle are being installed twice for now as it's not causing any issue in practice. So we should be able to merge this PR with the regular `requirements-dev.txt` and then rework this somehow.

Please see https://github.com/kernelci/kernelci-api/pull/218

**Comment by gctucker** [comment]
> > I think we can ignore the fact that pylint and pycodestyle are being installed twice for now as it's not causing any issue in practice. So we should be able to merge this PR with the regular `requirements-dev.txt` and then rework this somehow.
> 
> Please see #218

Yes, I mean we don't need to depend on #218 to be merged before merging this one.

File: .github/workflows/main.yml (hunk 7 above)
Comment (gctucker) [reply .github/workflows/main.yml:79]: Are you still experimenting with this?

File: .github/workflows/main.yml (hunk 6 above)
Comment (JenySadadia) [reply .github/workflows/main.yml]: I think `CHANGED_FILES` will be dependent on how many commits have been fetched.
I am experimenting with it by providing the `fetch-depth: 2`. But it seems like it doesn't change anything.

File: .github/workflows/main.yml (hunk 7 above)
Comment (JenySadadia) [reply .github/workflows/main.yml:78]: Updated the commit message.

File: .github/workflows/main.yml (hunk 6 above)
Comment (gctucker) [reply .github/workflows/main.yml]: It won't change anything if the PR has no more than 2 commits on top of the destination branch.  But most PRs have more than that, and rarely more than 20 so that's why I picked 32 as a "round" number.  If some commits are not checked out, then they won't have annotations.

Comment (JenySadadia) [reply .github/workflows/main.yml]: So, as an experiment, I tried `fetch-depth: 2` and `fetch-depth: 3` in https://github.com/kernelci/kernelci-api/pull/217. Instead of checking out commits based on provided value, it gets all the commits in the checkout and shows annotations for all modified files.

Comment (gctucker) [reply .github/workflows/main.yml]: OK yes so it is _also_ related to `CHANGED_FILES` as if the depth is less than the number of commits in the PR then the diff revision is not valid.

As we found out today, `--depth` is the number of extra commits to fetch in the history.  So with `--depth=1` you get the current revision and one history commit so 2 commits.

Comment (JenySadadia) [reply .github/workflows/main.yml:71]: > As we found out today, `--depth` is the number of extra commits to fetch in the history. So with `--depth=1` you get the current revision and one history commit so 2 commits.

Actually, that is true for `--depth=2`.
Also, I added another linter error to https://github.com/kernelci/kernelci-api/pull/217 to confirm if the annotations are shown based on the provided depth. It is working fine.

**Comment by JenySadadia** [comment]
Could you please take another look at this @gctucker?
I think it is ready to be merged.

File: .github/workflows/main.yml (hunk 6 above)
Comment (gctucker) [reply .github/workflows/main.yml:71]: I'm pretty sure there were 3 commits with `--depth=2` but maybe that's because one was a merge commit.  Either way, I think that's resolved now.

**Review by gctucker — APPROVED** [review]
